### PR TITLE
Add optional `IncludeDetectorPosition` to `CreateDetectorTable`

### LINKED ISSUE #38958: Add option to include detector position column in `CreateDetectorTable`
**Is your feature request related to a problem? Please describe.**
If you are using `CreateDetectorTable` it is quite possible you are also interested in the position of the detectors for each spectra. Would be nice to have an additional flag that will allow you to include this information

**Describe the solution you'd like**
Optional True/False IncludeDetectorPosition that, if True, will append the detector positions to the table

### PR DESCRIPTION
**Pull request by andy-bridger**
#### Summary of work
Optional True/False `IncludeDetectorPosition` that, if `True`, will append the detector positions to the `DetectorTable`

Fixes #38958. 

#### Further detail of work
Added the option, should default to False so shouldn't break any scripts and appends to the end so should have minimal impact even when columns referenced by index.

Added a test that the additional column is present with the flag included

### To test:

You can run the script:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws1 = Load("ENGINX00193749.nxs")

ws1_detectors_no_pos = CreateDetectorTable("ws1")                        
ws1_detectors = CreateDetectorTable("ws1", IncludeDetectorPosition = True)

group_ws = CreateGroupingWorkspace(InstrumentName='ENGINX', 
                        ComponentName='ENGIN-X', 
                        CustomGroupingString='100001-100100,101001-101100, 102001-102100')

ws2 = GroupDetectors(InputWorkspace="ws1", CopyGroupingFromWorkspace= "group_ws")
                     
ws2_detectors = CreateDetectorTable("ws2", IncludeDetectorPosition = True)
```

The file is in the System Test Data (`build\ExternalData\Testing\Data\SystemTest\ENGINX00193749.nxs`). This script should produce three `DetectorTables`, the first two should be identical except for the extra `Position` column. `ws2_detectors` should have only 3 spectra with the grouping defined in `group_ws` applied, each should have a single position. 

*The functionality of the presence of the position column is covered by the unit tests added, and the actual population of this column should be covered by the tests for the `spectrumInfo.position()` as that is what is being called.*

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by thomashampson** [comment]
I think your release notes want to go into framework/algorithms/new_features

File: Framework/Algorithms/src/CreateDetectorTable.cpp [hunk 1]
```diff
@@ -317,6 +331,9 @@ void populateTable(ITableWorkspace_sptr &t, const MatrixWorkspace_sptr &ws, cons
       if (includeDiffConstants) {
         colValues << 0.0 << 0.0 << 0.0 << 0.0;
       }
+      if (includeDetectorPosition) {
+        colValues << V3D(0.0, 0.0, 0.0);
```
Comment (RichardWaiteSTFC) [inline Framework/Algorithms/src/CreateDetectorTable.cpp:335]: I don't think the test you added will hit this branch will it? 
I've manually tested this using `PEARL00112777.nxs` (e.g. spectrum index 5) - so it works! But goof practice to try and cover every line you add with a test (not always easy though).

File: docs/source/release/v6.13.0/Framework/Algorithms/New_features/38960.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+- :ref:`CreateDetectorTable<algm-CreateDetectorTable>` now has an optional argument, ``IncludeDetectorPosition``, which, if `True`, will add a column, ``Position``, to the end of the ``DetectorTable``. This column gives the absolute detector position associated with each spectra in the table.
```
Comment (RichardWaiteSTFC) [inline docs/source/release/v6.13.0/Framework/Algorithms/New_features/38960.rst:1]: Not sure what `absolute detector position`  meant, would this also be accurate?
```suggestion
- :ref:`CreateDetectorTable<algm-CreateDetectorTable>` now has an optional argument, ``IncludeDetectorPosition``, which, if `True`, will add a column, ``Position``, to the end of the ``DetectorTable``. This column has the detector position as a ``V3D`` type object for each spectrum in the table.
```

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Changes look good, thanks for adding this feature and a test!

I note that if you `right-click on a workspace > Show Detectors` this calls `CreateDetectorTable` with the default `IncludeDetectorPosition=False` I think this is probably safest.

I just wanted to check whether you needed an additional test to hit the branch which puts in a null position `(0,0,0)`.

Also I have tested this on a peaks workspace (I didn't know this was possible until looking at the tests!) and the new option doesn't do anything (only 2 columns Index and DetectorID are printed for peak tables). I wonder if it would be possible to disable the new option if a peaks workspace is selected? if that's too difficult we would normally at 
 least print a warning saying it is ignored.

File: docs/source/release/v6.13.0/Framework/Algorithms/New_features/38960.rst (hunk 2 above)
Comment (andy-bridger) [reply docs/source/release/v6.13.0/Framework/Algorithms/New_features/38960.rst:1]: It is just what the doc string of `spectrumInfo().position()` said (`Returns the absolute position of the spectrum with the given index.`) and I didn't want to pass comment about how it was calculated - very happy to make the change, however

Comment (RichardWaiteSTFC) [reply docs/source/release/v6.13.0/Framework/Algorithms/New_features/38960.rst:1]: Fair enough `absolute position` seems to be used quite a lot in the repo...it looks to me like this is the average position of all the detectors contributing to the spectrum
https://github.com/mantidproject/mantid/blob/979f1465b595e6a41fac0879cb78365b61e4703e/Framework/API/src/SpectrumInfo.cpp#L137-L144
I guess normally `absolute` would refer to the magnitude but it isn't in this case...

**Review by RichardWaiteSTFC — DISMISSED** [review]
Great, thanks for adding the tests (as you pointed out there was no test coverage for the null branch so that's really nice to have)!

File: Framework/Algorithms/test/CreateDetectorTableTest.h [hunk 3]
```diff
@@ -86,6 +89,38 @@ class CreateDetectorTableTest : public CxxTest::TestSuite {
...
+  void test_Exec_Matrix_Workspace_with_no_valid_spectra() {
+    Workspace2D_sptr inputWS = WorkspaceCreationHelper::create2DWorkspaceWithFullInstrument(1, 10);
+    auto &spec = inputWS->getSpectrum(0);
+    spec.clearDetectorIDs(); // clear the detectors to test the exception catching
+
+    CreateDetectorTable alg;
+    TS_ASSERT_THROWS_NOTHING(alg.initialize())
+    TS_ASSERT(alg.isInitialized())
+    TS_ASSERT_THROWS_NOTHING(alg.setProperty("InputWorkspace", std::dynamic_pointer_cast<MatrixWorkspace>(inputWS)));
+    TS_ASSERT_THROWS_NOTHING(alg.execute();)
```
Comment (jclarkeSTFC) [inline Framework/Algorithms/test/CreateDetectorTableTest.h:101]:
```suggestion
    TS_ASSERT_THROWS_NOTHING(alg.execute());
```
Did you mean this instead?

File: Framework/Algorithms/test/CreateDetectorTableTest.h [hunk 4]
```diff
@@ -164,6 +199,70 @@ class CreateDetectorTableTest : public CxxTest::TestSuite {
...
+    Workspace2D_sptr inputWS = WorkspaceCreationHelper::create2DWorkspaceWithFullInstrument(1, 10);
+    auto &spec = inputWS->getSpectrum(0);
+    spec.clearDetectorIDs(); // clear the detectors to test the exception catching
+
+    CreateDetectorTable alg;
+    TS_ASSERT_THROWS_NOTHING(alg.initialize())
+    TS_ASSERT(alg.isInitialized())
+    TS_ASSERT_THROWS_NOTHING(alg.setProperty("InputWorkspace", std::dynamic_pointer_cast<MatrixWorkspace>(inputWS)));
+    TS_ASSERT_THROWS_NOTHING(alg.setProperty("IncludeDetectorPosition", true));
+    TS_ASSERT_THROWS_NOTHING(alg.execute();)
```
Comment (jclarkeSTFC) [inline Framework/Algorithms/test/CreateDetectorTableTest.h:243]:
```suggestion
    TS_ASSERT_THROWS_NOTHING(alg.execute());
```
Same as above?

File: Framework/Algorithms/test/CreateDetectorTableTest.h [hunk 5]
```diff
@@ -164,6 +199,70 @@ class CreateDetectorTableTest : public CxxTest::TestSuite {
...
+        ws = AnalysisDataService::Instance().retrieveWS<TableWorkspace>(inputWS->getName() + "-Detectors"));
+    TS_ASSERT(ws);
+
+    if (!ws) {
+      return;
+    }
+
+    // Check the results
+    TS_ASSERT_EQUALS(ws->columnCount(), 12); // extra column compared to test_Exec_Matrix_Workspace
+    TS_ASSERT_EQUALS(ws->rowCount(), 2);
```
Comment (jclarkeSTFC) [inline Framework/Algorithms/test/CreateDetectorTableTest.h:227]: Could this test also check the cell content, like the test when there are no valid detectors?

**Review by jclarkeSTFC — CHANGES_REQUESTED** [review]
Works as described, tests are there too. I spotted a couple of potential typos in the tests, and was wondering if one of the tests could have an extra check (should be a one-liner I think).

File: Framework/Algorithms/test/CreateDetectorTableTest.h (hunk 3 above)
Comment (jclarkeSTFC) [reply Framework/Algorithms/test/CreateDetectorTableTest.h:101]: I didn't spot that, yes I'd put the semicolons on the `TS_ASSERT`, etc statements as well, right at the end of the line.

File: Framework/Algorithms/test/CreateDetectorTableTest.h (hunk 5 above)
Comment (andy-bridger) [reply Framework/Algorithms/test/CreateDetectorTableTest.h:227]: I realise that testing the table isn't unintentionally coming from that exception is actually quite valuable, so I'll add that to all the applicable tests

**Review by jclarkeSTFC — APPROVED** [review]
Thanks for making those changes
